web/src/components/lineage/Lineage.tsx

```diff
@@ -143,7 +143,16 @@ class Lineage extends React.Component<LineageProps, LineageState> {
   getSelectedPaths = () => {
     const paths = [] as Array<[string, string]>
 
+    // Sets used to detect cycles and break out of the recursive loop
+    const visitedNodes = {
+      successors: new Set(),
+      predecessors: new Set()
+    }
+
     const getSuccessors = (node: string) => {
+      if (visitedNodes.successors.has(node)) return
+      visitedNodes.successors.add(node)
+
```

Comment:
The code looks good here. Would you mind looking into a test case to isolate this problem with a little graph instance. I'd also like to improve upon the reliance on the global graph object defined in this component. I think this could take in a graph as a parameter instead of extracting it from a higher place. Let me know what you think and thanks for taking the helm on this!

Reply:
Are there any existing tests for `Lineage.tsx` that I could build on? I thought of adding one but wasn't familiar enough with the codebase to know how to stub things in. Agree that it's probably cleaner / more testable if we make `getSelectedPaths` more pure by accepting these things as params. Maybe it makes sense to do that for all methods of `Lineage` in a separate PR.

Reply:
Ah looks like I can add one pretty easily in `web/src/__tests__/components/` I'll take a look at that later.

Reply:
Thanks, and yes it should be simple enough to initialize a dummy graph object there and test out this cycle issue.